packages/vite/src/node/server/shortcuts.ts

```diff
@@ -0,0 +1,81 @@
+import colors from 'picocolors'
+import type { ViteDevServer } from '..'
+import { openBrowser, resolveBrowserUrl } from './openBrowser'
+
+export function bindShortcuts(server: ViteDevServer): void {
+  if (!server.httpServer) return
+
+  server.config.logger.info(
+    `  > Shortcuts: ` +
+      SHORTCUTS.map((shortcut) => {
+        return `"${shortcut.key}" ${shortcut.name}`
+      }).join(', ')
+  )
+
+  const onInput = (input: string) => {
+    // ctrl+c or ctrl+d
+    if (input === '\x03' || input === '\x04') {
+      return process.kill(process.pid, 'SIGINT')
+    }
+    const shortcut = SHORTCUTS.find((shortcut) => shortcut.key === input)
+    shortcut?.action(server)
+  }
+
+  process.stdin
+    .on('data', onInput)
+    .setEncoding('utf8')
+    .setRawMode(true)
+    .resume()
+
+  server.httpServer.on('close', () => {
+    process.stdin.off('data', onInput).pause()
+  })
+}
+
+export interface Shortcut {
+  key: string
+  name: string
+  action(server: ViteDevServer): void | Promise<void>
+}
+
+export const SHORTCUTS: Shortcut[] = [
+  {
+    key: 'r',
+    name: 'restart',
+    action(server: ViteDevServer): void {
+      server.restart()
+    }
+  },
+  {
+    key: 'o',
+    name: 'open browser',
+    async action(server: ViteDevServer): Promise<void> {
+      const url = await resolveBrowserUrl(server)
+      openBrowser(url, true, server.config.logger)
+    }
+  },
+  {
+    key: 'f',
+    name: 'force restart',
```

Comment:
Sounds good to me!

@antfu @patak-dev  Any opinion here?

Reply:
I think we shouldn't expose this option. It is only needed if there is a bug in dependencies optimization re-bundling logic, that at this point is quite stable. I think it may end up confusing more users than helping them.

Reply:
For posterity, I'd also suggest changing the `key` to some variation of "r" (maybe "R") if this option makes a comeback in the future.